### ISSUE TITLE
I would like to learn from You how to run a function in a separate thread?

### ISSUE THREAD

**Issue by egor230** [ThePhD/sol2#853]
Thank You for such a wonderful multifunctional library. I would like to learn from You how to run a function in a separate thread? I'll try to explain it more simply: there is a main function, there is an infinite loop, everything is fine, it can run a function that runs a separate function in another thread, and there is work in multithreading. How to implement it in sql? I really need it in my lua plugin for gta vc. Thank You in advance for your reply.

```
function funs(ped, text)
2.	 player = findplayer()-- получить игрока
3.	 x,y,z =getcoordinates_on_y(player,5)-- получить координаты на 5 м впереди.
4.	 man = Createped(ped,x,y,z) -- создать педа.
5.	 mar = create_marker_actor(man) -- создать маркер педа.
6.	while true do 
7.	 if 0 == gethealth(man)-- проверить его здоровье.
8.	 then  
9.	 printmessage(text, 1500,1)
10.	 remove_ped(man)-- удалить педа.
11.	 removemarker(mar)-- удалить маркер над педом.
12.	 break
13.	 end
14.	end
15.	end
16.	function main()
17.	while lualoader == nil do
18.	 wait()
19.	 player = findplayer()-- получить игрока
20.	 if Keypress(VK_G)
21.	 then 
22.	 newthread(funs,"BMYBB","remove ped")-- функция, модель пед, текст надписи.
23.	 end 
24.	 
25.	 end  
26.	 end

```
tried so, did not work.
```
using namespace sol; 
const char* LUA = R"(
 function foo ()
print(223)
 end
 function f()
print("func f")
 end
)";
int starfunc(state lua1) {
	state_view lua(lua1); // non-owning

//	state lua;// Lua состояние.
	//sol::thread run = sol::thread::create(lua.lua_state());
//	state_view runstate = run.state();
	lua.open_libraries(lib::base, lib::package, lib::coroutine); // открыть доп.библиотеки.
	//runstate.script(LUA1);
	lua.script(LUA);// Вызвать функцию и получить результат.
	lua["foo"]();
	return 0;
};
int main(int argc, char* argv[]) {
	state lua;// Lua состояние.
	lua.open_libraries(lib::base, lib::package); // открыть доп.библиотеки.
	std::thread th(starfunc, lua);
	th.detach();
	lua.script(LUA);// Сначала сохранить результат в переменной, затем вызвать.
	sol::function f = lua["f"];
	f();

	return 0;
};

```

**Comment by egor230** [comment]
так тоже не выходит.
```
const char* LUA = R"(
 function foo ()
while true do
print(223)
 end
end 
 function f()
print("func f")
 end
)";
int starfunc(lua_State* L) {
	state_view lua(L); // non-owning

	lua.open_libraries(lib::base, lib::package, lib::coroutine); // открыть доп.библиотеки.
	lua.script(LUA);// Вызвать функцию и получить результат.
	lua["foo"]();
	return 0;
};
int main(int argc, char* argv[]) {
	lua_State* L = luaL_newstate();/*Функция создает новое Lua состояние. Она вызывает lua_newstate с функцией-*/
	luaL_openlibs(L);

	state lua;// Lua состояние.
	lua.open_libraries(lib::base, lib::package); // открыть доп.библиотеки.
	std::thread th(starfunc, L);
	th.detach();
	lua.script(LUA);// Сначала сохранить результат в переменной, затем вызвать.
	sol::function f = lua["f"];
	f();

	return 0;
};

```

**Comment by OrfeasZ** [comment]
Lua is **not** thread safe. Doing operations on a lua state (VM) from across multiple threads at the same time will lead in corruption and will almost definitely crash your app. You might want to look into `sol::thread` ([here](https://sol2.readthedocs.io/en/latest/api/thread.html)) and `sol::coroutine` ([here](https://sol2.readthedocs.io/en/latest/api/coroutine.html)). Keep in mind that these are not the same thing as CPU threads and behave differently.

**Comment by ThePhD** [comment]
What @OrfeasZ said. You need to restructure your code so that the Lua state is either accessed with a mutex + lock, or have a separate state per thread. There are a lot of ways to do this and unfortunately I can't just design the solution for you, because I don't know _exactly_ what your needs are.

I would really suggest having 2 different Lua states, and then explicitly serializing between the two when you need to share between threads. That is, take the data from one `sol::state`, put it into a C++ value representation outside of the `sol::state`, and then explicitly transfer/set those values into the new thread's `sol::state`. You'll have to work a little bit at the serialization.

I may have to write an example of this since it's becoming an increasingly common request to serialize data between 2 threads. I'll put it on my TODO list.

**Comment by ThePhD** [comment]
New example showcasing worker and basic function serialization with processing: https://github.com/ThePhD/sol2/blob/develop/examples/source/docs/std_thread.cpp

Minor update to threading page: https://sol2.readthedocs.io/en/latest/threading.html

**Comment by egor230** [comment]
> Новый пример, демонстрирующий рабочий и базовую сериализацию функций с обработкой: https://github.com/ThePhD/sol2/blob/develop/examples/source/docs/std_thread.cpp
> 
> Незначительное обновление страницы потоков: [https://sol2.readthedocs.io/en/latest/threading.html.](https://sol2.readthedocs.io/en/latest/threading.html)

thank you very much for Your reply. I apologize for the fact that I can only answer now.
I try to implement multithreading in Lua Api through coroutines using lua_sethook.
here's the code
```

void LUAHook(lua_State* L, lua_Debug* ar) {
    lua_yield(L, 0);
};
 const char* LUA = R"(
function foo()
for i = 1, 4 do
print(" func foo \n")
end end

function main()
for i = 1, 3 do
print(" func main "..i.."\n")
if j == nil
then
 foo1()
j=1
end
end  end
)";
int foo1(lua_State* L) {
    lua_State* L1 = lua_newthread(L); 
    lua_sethook(L, LUAHook, LUA_MASKLINE, 1);
    lua_getglobal(L1, "foo");
    lua_sethook(L1, LUAHook, LUA_MASKLINE, 1);
    lua_resume(L1, L, 0);
return  0;
};
int main() {
    lua_State* L = luaL_newstate(); luaL_openlibs(L);
    lua_register(L, "foo1", foo1);
    luaL_dostring(L, LUA);
    int ret;
    lua_getglobal(L, "main");// функция lua возобновляется с последней 
    //  прерванной позиции.
    ret = lua_resume(L, NULL, 0);


    return 0;
};
```
me need to display the following in the console

func main 1
func foo 
func main 2 
func foo 
func main 3 
func foo
